### ISSUE TITLE
Mobile: Caret moves to beginning on cursor change

### ISSUE THREAD

**Issue by raineorshine** [cybersemics/em#981]
## Steps to Reproduce

1. Create thought `a`
1. Create subthought `foo bar`
1. Set the cursor on `a`
1. Tap in between `foo` and `bar`

## Current behavior

The caret goes to the beginning of `foo bar`. (This does not occur when moving between siblings.)

If you look closely, you can see that the caret moves to the correct position for a split second then gets reset to the beginning. 

## Expected Behavior

The caret should move to in between `foo` and `bar`, where the user tapped. 

When the user taps on a thought and the cursor changes from parent to child, or child to parent, the caret should move to the point that the user tapped.

**Comment by onurpolattimur** [comment]
Here is what I have found so far. 

The Editable component is rendered twice after we tap in between foo and bar. After the second render, the caret position is set the zero. Because ` { offset: cursorOffset || state.cursorOffset || 0 }` is always zero. That is why caret at the beginning.

It is after the second render because `isEditing` is false and so, it doesn't enter the if block after the first render.
https://github.com/cybersemics/em/blob/fdd917169a202b537bd987ac858ccbbb8747c2fc/src/components/Editable.tsx#L367-L389

If you block the second render, you will see it works. I blocked the second render by adding the code below(just for debug) at the end of the `Editable.tsx`.

```
export default connect()(React.memo(Editable, (prev: any, next: any) => {
  console.log(prev, next)
  return !prev.isEditing && next.isEditing
}))

```

Same issue occurs on desktop if we add some delay before calling `setSelectionToCursorOffset`

I think we only need ` setSelectionToCursorOffset()` for newly created thoughts. I am not sure but maybe we can add a local state like 'isFirstRender' and we can set true this in useEffect. I tried and I can not see any problem.

**Comment by raineorshine** [comment]
@onurpolattimur Thank you! That is super helpful information.

Before considering blocking the second render, why is it rendering twice to begin with? If we can remove the extra render we could fix the bug and improve performance.

**Comment by onurpolattimur** [comment]
I am so sorry, I think I was wrong. It is not rendered twice, the first render log was for editable element contains 'a'. 

But I think, I found the root cause. As I said before, we always set caret position to zero. We need to continue to set zero for newly created thoughts to set caret.

```jsx
{ offset: cursorOffset || state.cursorOffset || 0 }
```

But my solution idea is still valid. We can use useEffect and useState to know whether Editable component is newly added. For all existing Editable component instances this state will be true after first render. So, we wont't set the caret position after click an existing Editable component. But this state will be false for newly created thought's Editable and in first render we will set this state to true  for the new Editable component instance and also set the caret position to zero.

To sum up, we need to set caret position only for  Editable of newly created thought. If there is another way to understand the Editable component is being rendered for newly created thought, we can also use that.

**Comment by raineorshine** [comment]
> I am so sorry, I think I was wrong. It is not rendered twice, the first render log was for editable element contains 'a'.

That's good at least :).

> As I said before, we always set caret position to zero.

`state.cursorOffset` is sometimes set via the `offset` argument to `setCursor`, e.g.:

https://github.com/cybersemics/em/blob/9305b92385015fd34489a49dc66e123facd80b67/src/reducers/deleteEmptyThought.ts#L87-L92

> To sum up, we need to set caret position only for Editable of newly created thought. If there is another way to understand the Editable component is being rendered for newly created thought, we can also use that.

The caret needs to be set whenever the cursor moves, not just when a new thought is created. So unless I'm missing something, I think we need a more surgical fix.

**Comment by onurpolattimur** [comment]
Okay, now everything is more clear.

**Why do we set the caret?** 

On events that don't contains user click, we should set the caret for editable elements. Because the user doesn't click anywhere but we want the caret to be on an thought. Here are some examples of this situation; 
- When we refresh the page, the caret should be on the thought in the path.
- When we delete a thought, the caret should be on the previous thought.
- When we add a thought, the caret should be on the newly created thought.
- ....

**What should we do?**

We should not set the caret if we want to see that the caret on the correct position where the user clicks. Because the browser already set the caret where the user clicks. The only thing we should know before setting caret is whether the user clicked or not.
